ext/XS-APItest/t/throwing.t

```diff
@@ -0,0 +1,159 @@
+#!perl
+
+use v5.36;
+
+# Test::More doesn't have fresh_perl_is() yet
+
+BEGIN {
+    require '../../t/test.pl';
+};
+
+use XS::APItest;
+
+package RunOnDestruct {
+    sub DESTROY { $_[0]->[0]->() }
+}
+sub run_on_destruct :prototype(&) {
+    return bless [@_], "RunOnDestruct";
+}
+
+# PL_throwing quiescently
+{
+    my $throwing = XS::APItest::PL_throwing;
+    ok(defined $throwing, 'PL_throwing is defined initially');
+    ok(!$throwing,        'PL_throwing is false initially');
+}
+
+# PL_throwing during normal leave
+{
+    my $var;
+    {
+        my $guard = run_on_destruct { $var = 0 + XS::APItest::PL_throwing };
+    }
+    is($var, 0, 'var is false after normal scope leave');
+}
+
+# PL_throwing during normal eval leave
+{
+    my $var;
+    eval {
+        my $guard = run_on_destruct { $var = 0 + XS::APItest::PL_throwing };
+    };
+    is($var, 0, 'var is false after normal eval leave');
+    ok(!XS::APItest::PL_throwing, 'PL_throwing false after eval returned');
+}
+
+# PL_throwing during exceptional eval leave
+{
+    my $var;
+    eval {
+        my $guard = run_on_destruct { $var = 0 + XS::APItest::PL_throwing };
+        die "Oopsie";
+    }; # ignore the error
+    is($var, 1, 'var is true after die in eval');
+    ok(!XS::APItest::PL_throwing, 'PL_throwing false after eval returned');
+}
+
+# PL_throwing during string-eval()
+{
+    my $var;
+    eval q(
+        my $guard = run_on_destruct { $var = 0 + XS::APItest::PL_throwing };
+        die "Oopsie";
+    ); # ignore the error
+    is($var, 1, 'var is true after die in string-eval');
+    ok(!XS::APItest::PL_throwing, 'PL_throwing false after string-eval returned');
+}
+
+# It's (probably) not possible to observe PL_throwing=true during a BEGIN,
+# CHECK or INIT phaser, but we can test END
+fresh_perl_is(
+    <<'EOF',
+use strict;
+use XS::APItest;
+END { printf "PL_throwing=%s\n", XS::APItest::PL_throwing ? "true" : "false" }
+die "Oopsie\n";
+EOF
+    "Oopsie\nPL_throwing=true",
+    { stderr => 1 },
+    'PL_throwing during END block');
+
+# PL_throwing during normal defer
+{
+    use experimental 'defer';
+
+    my $var;
+    eval {
+        defer { $var = 0 + XS::APItest::PL_throwing; }
+    };
+    is($var, 0, 'var is true during normal defer');
+    ok(!XS::APItest::PL_throwing, 'PL_throwing false after eval');
+}
+
+# PL_throwing during exceptional defer
+{
+    use experimental 'defer';
+
+    my $var;
+    eval {
+        defer { $var = 0 + XS::APItest::PL_throwing; }
+        die "Oopsie";
+    }; # ignore the error
+    is($var, 1, 'var is true during exceptional defer');
+    ok(!XS::APItest::PL_throwing, 'PL_throwing false after eval');
+}
+
+# PL_throwing at top-level defer
+fresh_perl_is(
+    <<'EOF',
+use strict;
+use experimental 'defer';
+use XS::APItest;
+defer { printf "PL_throwing=%s\n", XS::APItest::PL_throwing ? "true" : "false" }
+die "Oopsie\n";
```

Comment:
I think you might want to make this into a loop and wrap these two lines in CHECK/UNITCHECK/BEGIN/INIT blocks and make sure that they all the do the right thing. I also think you might want to check nesting the blocks, Eg, `X { eval "Y { defer ...; die ..; }" ... } ...` and validate that they all work as you expect. We have found that these combinations can lead to weird results before. Another thing to check is that they work under the debugger. I would take a look at t/op/eval.t and t/comp/retainedlines.t for examples.  Also see the tests in https://github.com/Perl/perl5/pull/20398